crates/driver/src/domain/competition/solution/trade.rs

```diff
@@ -13,14 +20,72 @@ pub enum Trade {
 pub struct Fulfillment {
     pub order: competition::Order,
     /// The amount executed by this fulfillment. See
-    /// [`competition::order::Partial`]. If the order is not partial, the target
-    /// amount must equal the amount from the order.
+    /// [`competition::order::Partial`]. If the order is not partial, the
+    /// executed amount must equal the amount from the order.
     pub executed: competition::order::TargetAmount,
 }
 
+impl Fulfillment {
+    /// Calculate the settlement contract input and output amounts executed by
+    /// this trade.
+    pub fn execution(&self, clearing_prices: &ClearingPrices) -> Result<Execution, Error> {
+        let input = self.executed.to_asset(&self.order);
+        let (input_price, output_price) = {
+            let sell_price = clearing_prices
+                .0
+                .get(&self.order.sell.token)
+                .ok_or(Error::ClearingPriceMissing)?
+                .to_owned();
+            let buy_price = clearing_prices
+                .0
+                .get(&self.order.buy.token)
+                .ok_or(Error::ClearingPriceMissing)?
+                .to_owned();
+            match self.order.side {
+                order::Side::Buy => (buy_price, sell_price),
+                order::Side::Sell => (sell_price, buy_price),
+            }
```

Comment:
I think this match isn't needed. `sell` tokens always enter the contract, `buy` tokens always leave the contract (regardless of the `order.side`). The difference between the `order.side` is only in which of the token amounts are fixed.